### PR TITLE
Unexpected error handling

### LINKED ISSUE #171: Handling promises error properly when making queries to backend
**Describe the bug**
If a call is made to the backend, we are assuming (at someplace, maybe everywhere) that it will return data and not necessarily raise an error. 

One of my chrome extension blocked a call since it thought it was a tracker or something. The app was trying to make a call to https://sso-dev.ised-isde.canada.ca/auth/realms/individual/protocol/openid-connect/token and the promise return an error, which then doesn't load any page except / (which doesn't need authentication).

**To Reproduce**
Install the EEF Privacy Badger extension and login to the app. Once logged in, the error will be raised

**Expected behavior**
We should handle these errors and show related something to the user instead of a blank page.

**Screenshots**
image

**Additional context**
Probably related to not having a catch for the keycloak promise call

### PR DESCRIPTION
**Pull request by TrevorBivi**
Adds handling for network errors

- When most get requests fail they will redirect the user to an error page (This page can handle non network errors too but isn't used for this anywhere right now).

- The exception is if a network error on /secured/profile is because a profile does not exist in which case a message stating the user is on an invalid profile url will be displayed. 

- Unexpected put/post errors will display a toast notification that data failed to save.

errorThing

### CODE REVIEW

File: services/frontend-v3/src/pages/UnexpectedError.jsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  const styles = {
+    content: { textAlign: "left", maxWidth: "1200px", margin: "0px" },
+    errorTitle: {},
+    errorDescription: { marginBottom: "0.0em", marginLeft: "1em" },
+  };
+
+  return (
+    <Result
+      //status={resultProps.status}
```
Comment (dynamic11) [inline services/frontend-v3/src/pages/UnexpectedError.jsx]: remove commented code

**Review by dynamic11 — CHANGES_REQUESTED** [review]
please fix linting and i will look at it again

File: services/frontend-v3/src/redux/slices/errorsSlice.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        };
+      } else {
+        serializedError = {
+          title: error.message,
+          isAxiosError: false,
+          description: error.stack.split("\n"),
+        };
+      }
+
+      state = state.push(serializedError);
```
Comment (BenJeau) [inline services/frontend-v3/src/redux/slices/errorsSlice.js]: Something cool with redux toolkit, the state can be mutated, https://redux-toolkit.js.org/api/createReducer#direct-state-mutation,  and the line can be simplified by doing this
```suggestion
      state.push(serializedError);
```

**Review by BenJeau — COMMENTED** [review]
I just need to test it locally, but besides the comments I left, everything seems good!

**Comment by BenJeau** [comment]
Just tested it locally, I'm getting a React compilation error from the import of functions in ProfileCardsView.

Something I just saw in the screenshots, should we add a way for the user to go back to a "safe place" like home (the user could be stuck at the error page)?

File: services/frontend-v3/src/pages/UnexpectedError.jsx (hunk 1 above)
Comment (TrevorBivi) [reply services/frontend-v3/src/pages/UnexpectedError.jsx]: Completed

File: services/frontend-v3/src/components/admin/categoryTable/CategoryTable.jsx [hunk 3]
```diff
@@ -121,12 +123,12 @@ function CategoryTable({ intl, type }) {
     } catch (error) {
       // eslint-disable-next-line no-console
       console.log(error);
```
[the remark this answers is not in the document]
Comment (TrevorBivi) [reply services/frontend-v3/src/components/admin/categoryTable/CategoryTable.jsx]: Completed

File: services/frontend-v3/src/components/admin/schoolTable/SchoolTable.jsx [hunk 4]
```diff
@@ -35,26 +36,27 @@ const SchoolTable = ({ type, intl }) => {
...
   useEffect(() => {
-    let schools = [];
     if (loading) {
       const setState = async () => {
-        schools = await getSchools();
-        setData(schools);
+        await getSchools()
+          .then(schools => setData(schools))
+          .catch(error => handleError(error, "redirect"));
         // eslint-disable-next-line no-console
```
[the remark this answers is not in the document]
Comment (TrevorBivi) [reply services/frontend-v3/src/components/admin/schoolTable/SchoolTable.jsx]: Completed

File: services/frontend-v3/src/components/admin/competencyTable/CompetencyTable.jsx [hunk 5]
```diff
@@ -85,13 +88,13 @@ const CompetencyTable = ({ intl, type }) => {
 
   /* handles reset of column search functionality */
   // Consult: function taken from Ant Design table components (updated to functional)
-  const handleReset = (clearFilters) => {
+  const handleReset = clearFilters => {
```
[the remark this answers is not in the document]
Comment (TrevorBivi) [reply services/frontend-v3/src/components/admin/competencyTable/CompetencyTable.jsx]: fixed

File: services/frontend-v3/src/history.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import { createBrowserHistory } from "history";
+
+const history = createBrowserHistory();
+export default history;
```
[the remark this answers is not in the document]
Comment (TrevorBivi) [reply services/frontend-v3/src/history.js:4]: I have made the changes discussed on MS teams

File: services/frontend-v3/src/components/profileCards/ProfileCardsView.jsx [hunk 7]
```diff
@@ -41,7 +42,10 @@ function ProfileCardsView({
     try {
       // Get current card visibility status from db
       const url = `${backendAddress}api/profile/${urlID}`;
-      const result = await axios.get(url);
+      const result = await axios.get(url).catch(error => {
+        handleError(error, "redirect");
+        return undefined;
```
[the remark this answers is not in the document]
Comment (TrevorBivi) [reply services/frontend-v3/src/components/profileCards/ProfileCardsView.jsx]: Removed unnecessary return

**Comment by TrevorBivi** [comment]
> Just tested it locally, I'm getting a React compilation error from the import of functions in ProfileCardsView.
> 
> Something I just saw in the screenshots, should we add a way for the user to go back to a "safe place" like home (the user could be stuck at the error page)?

I added this feature. /secured/home isn't a particularly safe place since it makes a few get requests, so I made a back to landing button instead

backToLanding

**Comment by BenJeau** [comment]
I cleared my db, and when 
szoter_annotated_image (1)
I get this and can't seem to setup my profile
image

**Comment by TrevorBivi** [comment]
> I cleared my db, and when
> I get this and can't seem to setup my profile

Whoops. I fixed that issue and the formatting problem I could see in your error photo, thanks!

**Review by BenJeau — APPROVED** [review]
Seems to work!

File: yarn.lock [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
```
Comment (BenJeau) [inline yarn.lock]: Should prob remove this file

File: services/frontend-v3/src/i18n/en_CA.json [hunk 9]
```diff
@@ -341,7 +343,13 @@
...
   "home": "Home",
 
-  "error.button": "Back to home",
+  "error.redirect.title": "Unexpected error",
+  "error.redirect.subtitle": "Something went wrong",
+  "error.button": "Back to Home",
+  "error.retry": "Retry",
+  "error.save": "Data not saved due to network error",
+  "error.show": "Show error log",
+  "error.hide": "Hide error log",
```
Comment (BenJeau) [inline services/frontend-v3/src/i18n/en_CA.json:358]: Should add all of these variables to fr_CA also

**Review by BenJeau — CHANGES_REQUESTED** [review]
A couple few last things

Upon clearing my database, I was at my profile page, then when I reloaded, it showed me this, instead of just redirecting me to the create profile flow
image

**Comment by dynamic11** [comment]
please add spacing between the buttons. You can use antd to do the auto spacing:
https://ant.design/components/space/

File: services/frontend-v3/src/i18n/en_CA.json (hunk 9 above)
Comment (TrevorBivi) [reply services/frontend-v3/src/i18n/en_CA.json:358]: done

File: services/frontend-v3/src/redux/index.js [hunk 10]
```diff
@@ -15,6 +15,7 @@ import rootReducer from "./slices";
 const persistConfig = {
   key: "root",
   storage,
+  whitelist: ["settings"],
```
Comment (BenJeau) [inline services/frontend-v3/src/redux/index.js]:
```suggestion
  whitelist: ["settings", "user"],
```

**Review by BenJeau — CHANGES_REQUESTED** [review]
Looks good, just one small thing, just add the user reducer to the whitelist in the persist config, it was acting weird in /admin. This should be my last request :p

**Review by BenJeau — APPROVED** [review]

**Review by dynamic11 — APPROVED** [review]
my backend keeps failing but im not sure if its an issue with me or the branch

**Pull request merged**